tests/integration/plugins/test_validate_environment.py

```diff
@@ -0,0 +1,281 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import subprocess
+import textwrap
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Set
+
+import pytest
+import yaml
+
+import craft_parts
+from craft_parts import Step, errors, plugins
+
+
+@pytest.fixture
+def mytool(new_dir):
+    tool = Path(new_dir, "mock_bin", "mytool")
+    tool.parent.mkdir(exist_ok=True)
+    tool.write_text("echo ok")
+    tool.chmod(0o755)
+    yield tool
+
+
+@pytest.fixture
+def mytool_not_ok(new_dir):
+    tool = Path(new_dir, "mock_bin", "mytool")
+    tool.parent.mkdir(exist_ok=True)
+    tool.write_text("echo not ok")
+    tool.chmod(0o755)
+    yield tool
+
+
+@pytest.fixture
+def mytool_error(new_dir):
+    tool = Path(new_dir, "mock_bin", "mytool")
+    tool.parent.mkdir(exist_ok=True)
+    tool.write_text("exit 22")
+    tool.chmod(0o755)
+    yield tool
+
+
+class AppPluginProperties(plugins.PluginProperties, plugins.PluginModel):
+    """The application-defined plugin properties."""
+
+    @classmethod
+    def unmarshal(cls, data: Dict[str, Any]):
+        return cls()
+
+
+class AppPluginEnvironmentValidator(plugins.PluginEnvironmentValidator):
+    """Check the execution environment for the app plugin."""
+
+    def validate_environment(self, *, part_dependencies: Optional[List[str]] = None):
+        """Ensure the environment contains dependencies needed by the plugin.
+
+        If mytool is created by a part, that part must be named `mytool`.
+        """
+        try:
+            output = self._execute("mytool")
+            if output.strip() != "ok":
+                raise errors.PluginEnvironmentValidationError(
+                    part_name=self._part_name,
+                    reason="mytool is expected to print ok",
+                )
+        except subprocess.CalledProcessError as err:
+            if err.returncode != 127:
```

Comment:
code comment for 127 please

Reply:
I'll use a constant to replace the magic number.